adafruit_bno080/uart.py

```diff
@@ -0,0 +1,161 @@
+        self._data_buffer[4:4+data_length] = data
+        self._uart.write(b'\x7e') # start byte
+        time.sleep(0.001)
+        self._uart.write(b'\x01') # SHTP byte
+        time.sleep(0.001)
+        for b in self._data_buffer[0:write_length]:
+            ba[0] = b
+            self._uart.write(ba)
+            time.sleep(0.001)
+        time.sleep(0.001)
+        self._uart.write(b'\x7e') # end byte
+
+        #print("Sending", [hex(x) for x in self._data_buffer[0:write_length]])
+        
+        self._sequence_number[channel] = (self._sequence_number[channel] + 1) % 256
+        return self._sequence_number[channel]
+
+
+    def _read_into(self, buf, start=0, end=None):
+        if end is None:
+            end = len(buf)
+
+        #print("Avail:", self._uart.in_waiting, "need", end-start)
+        for idx in range(start, end):
+            data = self._uart.read(1)
+            b = data[0]
+            if b == 0x7d: # control escape
+                data = self._uart.read(1)
+                b = data[0]
+                b ^=0x20
+            buf[idx] = b
+        #print("UART Read buffer: ", [hex(i) for i in buf[start:end]])
+
+    def _read_header(self):
+        """Reads the first 4 bytes available as a header"""
+        # try to read initial packet start byte
+        data = None
+        while True:
+            data = self._uart.read(1)
+            if not data:
+                continue
+            b = data[0]
+            if b == 0x7e:
+                break
+
+        # read protocol id
+        data = self._uart.read(1)
+        if data and data[0] == 0x7e:   # second 0x7e
+            data = self._uart.read(1)
+        if not data or data[0] != 0x01:
+            raise RuntimeError("Unhandled UART control SHTP protocol")
+        # read header
+        self._read_into(self._data_buffer, end=4)
+
+        #print("SHTP Header:", [hex(x) for x in self._data_buffer[0:4]])
+
+    def _read_packet(self):
+        self._read_header()
+        
+        #print([hex(x) for x in self._data_buffer[0:4]])
+        header = Packet.header_from_buffer(self._data_buffer)
+        packet_byte_count = header.packet_byte_count
+        channel_number = header.channel_number
+        sequence_number = header.sequence_number
+
+        self._sequence_number[channel_number] = sequence_number
+        if packet_byte_count == 0:
+            raise PacketError("No packet available")
+
+        self._dbg("channel %d has %d bytes available" % (channel_number, packet_byte_count-4))
+
+        if packet_byte_count > DATA_BUFFER_SIZE:
+            self._data_buffer = bytearray(packet_byte_count)
+
+        # skip 4 header bytes since they've already been read
+        self._read_into(self._data_buffer, start=4, end=packet_byte_count)
+
+        #print("Packet: ", [hex(i) for i in self._data_buffer[0:packet_byte_count]])
+        
+        data = self._uart.read(1)
+        b = data[0]
+        if b != 0x7e:
```

Comment:
@ladyada I'm hitting this after a few seconds with every run; I'm testing with a Metro M4 via RX/TX pins

Reply:
hitting what?

Reply:
@ladyada hitting the exception raise : `raise RuntimeError("Didn't find packet end")`